ebike_city_tools/optimize/round_optimized_sort_selection.py

```diff
@@ -0,0 +1,224 @@
+import math
+import pandas as pd
+import networkx as nx
+import numpy as np
+
+from ebike_city_tools.optimize.linear_program import define_IP
+from ebike_city_tools.optimize.round_simple import edge_to_source_target
+from ebike_city_tools.utils import (
+    lane_to_street_graph,
+    extend_od_circular,
+    compute_car_time,
+    compute_edgedependent_bike_time,
+    output_to_dataframe,
+)
+from ebike_city_tools.iterative_algorithms import transform_car_to_bike_edge
+from ebike_city_tools.metrics import compute_travel_times_in_graph
+
+FLOW_CONSTANT = 1
+
+def sort_by_bikevalue(capacities) :
+    """Sorts the capacities by planned bike capacity, with ties broken by car capacity."""
+    return capacities.sort_values(["u_b(e)", "u_c(e)"], ascending=[False, True])
+
+def rounding_error(fractional_value):
+    """Given a fractional value, returns the distance to the closest integer."""
+    next_int = round(fractional_value)
+    if fractional_value >= 0:
+        return 0
+    return abs(next_int - fractional_value)
+
+def rounding_error_of_row(row):
+    """Auxiliary funtion determining the rounding error of the car capacity."""
+    return rounding_error(row["u_c(e)"]) #+ rounding_error(row["u_c(e)_reversed"])
+
+def round_row(row):
+    """Given a row, this function returns the row obtained by rounding the car and bike capacities."""
+    # print(row)
+    new_value = min(1, round(row["u_c(e)"]))
+    # new_value = round(row["u_c(e)"])
+    rounded_by = row["u_c(e)"] - new_value
+    row["u_c(e)"] = new_value
+    row["u_b(e)"] = row["u_b(e)"] + rounded_by
+    return row
+
+def sort_by_rounding_error(capacities) :
+    """Sorts the capacities by increasing rounding error of the car capacity."""
+    cap_with_rounding_error = capacities
+    cap_with_rounding_error["rounding_error"] = capacities.apply(rounding_error_of_row, axis=1)
+    sorted_cap = cap_with_rounding_error.sort_values(["rounding_error", "u_b(e)"], ascending=[True, False])
+    sorted_cap = sorted_cap.apply(round_row, axis = 1)
+    return sorted_cap.drop(['rounding_error'], axis=1)
+
+class ParetoRoundOptimizeSortSelect:
+    def __init__(self, G_lane, od, rounding_method = "highest_bike_value", sp_method="od", optimize_every_x=5, **kwargs):
+        """
+        kwargs: Potential keyword arguments to be passed to the LP function
+        rounding_method: Specifies, how edges are selected for the rounding. Possible selections are:
+            - "highest_bike_value": Rounds up the highest bike value to a bike lane, if possible by connectivity.
+            - "lowest_rounding_error": Rounds the car value closest to an integer, if possible by connectivity.
+        """
+        self.od = extend_od_circular(od, list(G_lane.nodes()))
```

Comment:
This should probably be done in the `compare_rounding.py` script, to make sure that the same OD matrix (with extension) is used for all algorithms. Well spotted though that this was missing! My bad, I usually included it in the synthetic-data-generation part in each script, but seems like I forgot it here

Reply:
Ah, I got confused here - I added it to the Pareto Optimizer because the OD matrix should not be extended for the betweenness centrality algorithms (only for ours). So we should keep it as is.